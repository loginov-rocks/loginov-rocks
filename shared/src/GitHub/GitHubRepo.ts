export interface GitHubRepo {
  description: string;
  homepageUrl: string;
  isArchived: boolean;
  language: string;
  latestTag: string;
  name: string;
  stars: number;
  updatedAt: number;
  url: string;
}
