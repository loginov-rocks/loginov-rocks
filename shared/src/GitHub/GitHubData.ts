import { GitHubRepo } from './GitHubRepo';

export interface GitHubData {
  homepageUrl: string;
  login: string;
  repos: GitHubRepo[];
  url: string;
}
