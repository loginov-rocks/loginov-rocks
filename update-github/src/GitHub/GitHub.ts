import { GitHubData as Data, GitHubRepo as Repo } from '@loginov-rocks/loginov-rocks-shared';
import fetch, { Headers, Response } from 'node-fetch';

import { GitHubRepo } from './GitHubRepo';
import { GitHubTag } from './GitHubTag';
import { GitHubUser } from './GitHubUser';

interface Options {
  baseUrl: string;
  personalAccessToken: string;
}

export class GitHub {
  private readonly baseUrl: string;

  private readonly personalAccessToken: string;

  constructor({ baseUrl, personalAccessToken }: Options) {
    this.baseUrl = baseUrl;
    this.personalAccessToken = personalAccessToken;
  }

  private static parseNextPageUrl(headers: Headers): string | undefined {
    const header = headers.get('Link');

    if (!header) {
      return undefined;
    }

    const links = header.split(',');
    const link = links.find((l) => l.endsWith('>; rel="next"'));

    if (!link) {
      return undefined;
    }

    return link.trim().substring(1).replace('>; rel="next"', '');
  }

  private static async mergeArrayResponses<T>(responsesStack: Response[]): Promise<T> {
    const results = await Promise.all(responsesStack.map((response) => {
      if (!response.ok) {
        throw new Error('GitHub one of the requests in merge array responses stack failed');
      }

      return response.json();
    }));

    const [firstResult, ...otherResults] = results;

    if (otherResults) {
      return firstResult.concat(...otherResults);
    }

    return firstResult;
  }

  async getData(): Promise<Data> {
    const gitHubUser = await this.getGitHubUser();
    const repos = await this.getRepos();

    return {
      homepageUrl: gitHubUser.blog,
      login: gitHubUser.login,
      repos,
      timestamp: Date.now(),
      url: gitHubUser.html_url,
    };
  }

  /**
   * @see https://docs.github.com/en/rest/overview/resources-in-the-rest-api#oauth2-token-sent-in-a-header
   */
  private apiGet(url: string): Promise<Response> {
    const apiUrl = url.startsWith(this.baseUrl) ? url : `${this.baseUrl}${url}`;

    return fetch(apiUrl, {
      headers: {
        Authorization: `token ${this.personalAccessToken}`,
      },
    });
  }

  /**
   * @see https://docs.github.com/en/rest/overview/resources-in-the-rest-api#pagination
   */
  private async apiGetWithPagination(url: string, responsesStack: Response[] = []): Promise<Response[]> {
    const response = await this.apiGet(url);

    responsesStack.push(response);

    const nextPageUrl = GitHub.parseNextPageUrl(response.headers);

    if (nextPageUrl) {
      return this.apiGetWithPagination(nextPageUrl, responsesStack);
    }

    return responsesStack;
  }

  /**
   * @see https://docs.github.com/en/rest/reference/users#get-the-authenticated-user
   */
  private async getGitHubUser(): Promise<GitHubUser> {
    const response = await this.apiGet('/user');

    if (!response.ok) {
      throw new Error('GitHub get user request failed');
    }

    return response.json();
  }

  /**
   * @see https://docs.github.com/en/rest/reference/repos#list-repositories-for-the-authenticated-user
   */
  private async getGitHubRepos(): Promise<GitHubRepo[]> {
    const responsesStack = await this.apiGetWithPagination(
      '/user/repos?affiliation=owner&per_page=100&sort=full_name&visibility=public',
    );

    return GitHub.mergeArrayResponses<GitHubRepo[]>(responsesStack);
  }

  /**
   * @see https://docs.github.com/en/rest/reference/repos#list-repository-tags
   */
  private async getGitHubTags(ownerLogin: string, repoName: string): Promise<GitHubTag[]> {
    const response = await this.apiGet(`/repos/${ownerLogin}/${repoName}/tags`);

    if (!response.ok) {
      throw new Error('GitHub get tags request failed');
    }

    return response.json();
  }

  private async getRepos(): Promise<Repo[]> {
    const gitHubRepos = await this.getGitHubRepos();

    const repoToLatestTagMapping = await Promise.all(
      gitHubRepos.map(async (repo) => {
        const tags = await this.getGitHubTags(repo.owner.login, repo.name);

        return {
          latestTag: tags.length > 0 ? tags[0].name : '',
          repoName: repo.name,
        };
      }),
    );

    return gitHubRepos.map((repo) => {
      const repoToLatestTag = repoToLatestTagMapping.find(({ repoName }) => repoName === repo.name);
      const latestTag = repoToLatestTag ? repoToLatestTag.latestTag : '';

      return {
        description: repo.description || '',
        homepageUrl: repo.homepage || '',
        isArchived: repo.archived,
        language: repo.language || '',
        latestTag,
        name: repo.name,
        stars: repo.stargazers_count,
        updatedAt: new Date(repo.updated_at).getTime(),
        url: repo.html_url,
      };
    });
  }
}
