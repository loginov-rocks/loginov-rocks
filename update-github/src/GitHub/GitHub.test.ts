/* eslint-disable @typescript-eslint/no-var-requires, import/no-extraneous-dependencies */

import { GitHub } from './GitHub';

// eslint-disable-next-line global-require
jest.mock('node-fetch', () => require('fetch-mock-jest').sandbox());
const fetchMock = require('node-fetch');

const gitHubRepos = require('./__fixtures__/gitHubRepos.json');
const gitHubTagsMock = require('./__fixtures__/gitHubTags.json');
const gitHubUserMock = require('./__fixtures__/gitHubUser.json');

const gitHubReposMock = gitHubRepos.slice(0, 3);

const gitHub = new GitHub({
  baseUrl: 'https://api.github.com',
  personalAccessToken: 'personal-access-token',
});

describe('getData', () => {
  it('gets GitHub data using API', async () => {
    fetchMock.get('https://api.github.com/user', gitHubUserMock)
      .get(
        'https://api.github.com/user/repos?affiliation=owner&per_page=100&sort=full_name&visibility=public',
        gitHubReposMock,
      )
      .get('https://api.github.com/repos/loginov-rocks/Angular-Gulp-Boilerplate/tags', gitHubTagsMock)
      .get('https://api.github.com/repos/loginov-rocks/Anticloud/tags', gitHubTagsMock)
      .get('https://api.github.com/repos/loginov-rocks/anticloud-server/tags', gitHubTagsMock);

    const data = await gitHub.getData();

    fetchMock.reset();

    expect(data).toStrictEqual(
      expect.objectContaining({
        homepageUrl: 'https://loginov.rocks',
        login: 'loginov-rocks',
        url: 'https://github.com/loginov-rocks',
      }),
    );

    expect(Array.isArray(data.repos)).toBeTruthy();
    expect(data.repos.length).toBe(3);

    expect(data.repos[0]).toStrictEqual({
      description: 'Clean but full featured AngularJS boilerplate using Gulp workflow and best practices',
      homepageUrl: 'https://www.npmjs.com/package/angular-gulp-boilerplate',
      isArchived: false,
      language: 'JavaScript',
      latestTag: 'v1.0.10',
      name: 'Angular-Gulp-Boilerplate',
      stars: 29,
      updatedAt: 1614276648000,
      url: 'https://github.com/loginov-rocks/Angular-Gulp-Boilerplate',
    });

    expect(data.repos).toMatchSnapshot();
  });
});
