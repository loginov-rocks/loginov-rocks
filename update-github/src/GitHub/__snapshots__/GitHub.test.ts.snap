// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getData gets GitHub data using API 1`] = `
[
  {
    "description": "Clean but full featured AngularJS boilerplate using Gulp workflow and best practices",
    "homepageUrl": "https://www.npmjs.com/package/angular-gulp-boilerplate",
    "isArchived": false,
    "language": "JavaScript",
    "latestTag": "v1.0.10",
    "name": "Angular-Gulp-Boilerplate",
    "stars": 29,
    "updatedAt": 1614276648000,
    "url": "https://github.com/loginov-rocks/Angular-Gulp-Boilerplate",
  },
  {
    "description": "Training Repository",
    "homepageUrl": "",
    "isArchived": false,
    "language": "JavaScript",
    "latestTag": "v1.0.10",
    "name": "Anticloud",
    "stars": 0,
    "updatedAt": 1614278791000,
    "url": "https://github.com/loginov-rocks/Anticloud",
  },
  {
    "description": "Training Repository",
    "homepageUrl": "",
    "isArchived": false,
    "language": "JavaScript",
    "latestTag": "v1.0.10",
    "name": "anticloud-server",
    "stars": 0,
    "updatedAt": 1614276807000,
    "url": "https://github.com/loginov-rocks/anticloud-server",
  },
]
`;
